### PR TITLE
[core] Add uuid to expo-modules-core

### PR DESCRIPTION
**Pull request by alanjhughes**
# Why
Closes ENG-9174

# How
- Forked [uuid](https://github.com/uuidjs/uuid) @3.4.0 into `expo-modules-core`
- Removed the uuid dependency from expo
- Fixed all imports
- Ran tests in every package where it was being used

# Test Plan
Ran all tests and tested running bare-expo

### CODE REVIEW

File: packages/@expo/uuid/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+For bare React Native projects, you must ensure that you have [installed and configured the `expo` package](https://docs.expo.dev/bare/installing-expo-modules/) before continuing.
+
+### Add the package to your npm dependencies
+
+```
+npm install @expo/uuid
+```
+
+
+
```
Comment (brentvatne) [inline packages/@expo/uuid/README.md:25]:
```suggestion
```

File: packages/@expo/uuid/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# @expo/uuid
+
+Fork of UUID
```
Comment (brentvatne) [inline packages/@expo/uuid/README.md:3]:
```suggestion
Fork of `uuid@3.4.0` for usage without any dependency on native crypto APIs. This library is used with in several `expo` package. We do not intend for it to be used directly by developers. There may be breaking changes at any time — use `uuid` instead, if you are considering installing this.
```

**Review by brentvatne — CHANGES_REQUESTED** [review]

**Review by tsapeta — CHANGES_REQUESTED** [review]
Generally I think it would be better to put this package outside the monorepo. It's not going to change, even long-term I guess.
Especially that, as I said in many threads and discussions around this problem, this is not the right solution in my opinion. It's much easier and better to expose native UUID generation from `expo-modules-core` (similarly to `expo-crypto` with its `randomUUID`) instead of unnecessarily adding all those new files to the JS bundle.

In SDK 50 there will be a "core module" ([already implemented on iOS](https://github.com/expo/expo/blob/main/packages/expo-modules-core/ios/Swift/Modules/CoreModule.swift)) that decorates the `global.expo` object and will be the place for such things.
On web we can use `crypto.randomUUID()` instead, or if we need it for server-side rendering, use only a subset of `uuid` package that we actually need.

I understand that this might be urgent to fix for SDK 49, so I'll be fine with this PR if one of the following is met:
- The package is moved outside the repo.
- v4 algorithm is included within `expo-modules-core` package, but containing only necessary functionalities (no v35, v5, md5, sha1 algorithms). This will make us closer to what we will do in SDK 50 (we'll just switch the implementation to native on Android and iOS).

**Comment by brentvatne** [comment]
that sounds like a good plan @tsapeta!

File: react-native-lab/react-native
Comment (tsapeta) [inline react-native-lab/react-native:1]: Revert submodule update

**Review by ide — CHANGES_REQUESTED** [review]
The prior uuid module was deprecated because it used a weak RNG (Math.random()) to generate values, which is fine for our use cases, which are unrelated to security. However, it would be good to communicate that in the exported function names so no one uses the new uuid function for security-related use cases. Something like "weak" is maybe a good adjective to use. It should evoke light caution.

Also the function name should indicate that it generates UUID v4 values. @tsapeta pointed out places where we use UUID v5 for instance and we don't want to mix those up.

---

Separately, I'm wondering if we can both reduce the amount of JS (faster bundling times, smaller JS updates) and use a strong RNG by implementing UUID generation in Swift/C and Kotlin and just bridge a synchronous method through expo-modules-core. There are a couple of decisions we'd need to make like are we OK with adding a synchronous bridged method to e-m-c?

File: packages/expo/src/environment/getInstallationIdAsync.ts [hunk 3]
```diff
@@ -15,11 +14,12 @@ export default async function getInstallationIdAsync() {
   // It's unlikely identifierForVendor will be null (it returns null if the
   // device has been restarted but not yet unlocked), but let's handle this
   // case.
+
   if (identifierForVendor) {
-    installationId = uuidv5(`${bundleIdentifier}-${identifierForVendor}`, UUID_NAMESPACE);
+    installationId = uuid(`${bundleIdentifier}-${identifierForVendor}`);
```
[the remark this answers is not in the document]
Comment (brentvatne) [reply packages/expo/src/environment/getInstallationIdAsync.ts:19]: ah good catch. yeah we should keep using v5 here to be safe, i'm not actually familiar with why we chose to use it here though 🤔

**Comment by brentvatne** [comment]
lots of comments here are focused around the best thing for the future. the goal in this PR to solve an immediately problem that users often encounter when installing `uuid` alongside our copy of it in certain repo configurations, and this pr handles that without regressing behavior. i think we need to focus on what needs to be done to make this shippable and then create tasks to follow up in sdk 50 for future work.

**Comment by brentvatne** [comment]
- let's keep `@expo/uuid` in this repo for SDK 49, and we can remove it once we come up with another alternative
- let's revert the change to `getInstallationIdAsync` to ensure we use uuid v5 as before

it sounds like there are a bunch of ideas here for how to make things better in the future, but we are on a short timeline and i want to make the smallest possible change here that fixes the problems we're encountering and gets us in a better state.

**Comment by ide** [comment]
Deferring the possibility of native changes, I would like the new function to (1) make clear it's UUID v4 and (2) is weak and inappropriate for security ("weak"). Suggestion: `s/uuid()/weakUUIDv4()/`

**Comment by tsapeta** [comment]
@ide While I generally agree with clarifying its weakness and being v4, this is not going to be a part of the public API and we'll remove it in SDK 50 anyway. Not sure if you've read my previous comments, but we've already brought up that this needs to be implemented natively and this PR aims to just fix it temporarily to unblock people using SDK 49 from using different versions of uuid.

**Comment by ide** [comment]
@alanjhughes OK. I do think it is preferable to communicate the lack of safety of this function at least in a doc block explaining when and when not it is appropriate to use. In the future, it would be better for PRs like this to use the names of functions to communicate the right kind of hesitation people should have when using the functions.

**Comment by tsapeta** [comment]
@ide What do you say if we already shape it in a way that we would like it to be in SDK 50, but with an additional note in the doc comment that these are weak. In SDK 50 we will switch the implementation to native and remove that note.

I imagine the API could be like this 👇 similarly to the `uuid` package
```ts
export const uuid = {
  v4() { [native code] },
  v5(name, namespace) { [native code] }
}
```

**Comment by ide** [comment]
Probably better to have uuidV4() and uuidV5() instead of uuid.v4(). That way there aren't non-descriptive functions called "v4". It's much clearer to have a function named "uuidV4". Choose function names that make sense without extra context.

As for making the API change now or later, I don't think it matters that much. The important thing is to tell people not to use the function for anything where security or strong randomness is needed.

**Review by ide — APPROVED** [review]

**Review by tsapeta — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]

**Pull request merged**